File: python/RotateFunction.py
from typing import List


class Solution:
    def maxRotateFunction(self, A: List[int]) -> int:
        res, sumval, N, value = 0, sum(A), len(A), 0
        for i in range(N):
            value += i * A[i]
        res = value
        for i in range(1, N):
            value += sumval - N * A[N - i]
            res = max(res, value)
        return res

File: C/RotateFunction.c
#include <limits.h>

#define max(a, b) (((a) > (b)) ? (a) : (b))

int maxRotateFunction(int* A, int ASize)
{
	if (ASize == 0)
		return 0;
	long sum = 0, pre = 0, res = INT_MIN;
	for (int i = 0; i < ASize; ++i)
	{
		pre += (long)i * A[i];
		sum += A[i];
	}
	for (int i = ASize - 1; i >= 0; --i)
	{
		pre = pre + sum - (long)ASize * A[i];
		res = max(res, pre);
	}
	return res;
}
